[RFC] Target data file formats and organisation

annakrystalli opened the pull request:
This PR is a Request For Comment on an initial proposal for target data file formats and organisation

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+- new data can easily be added to the target data.(see [supplementary material](https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html)).
+
+- arrow datasets also allow us to filter the data on read which means we don't need to read the full dataset in should only a subset be required.
```
A suggestion for further decoupling specific tools from the the benefits of the proposal.

```suggestion
- partitioned datasets allow more efficient access by partition-aware data readers (for example, arrow, polars, DuckDB), which can filter the data on read if only a subset of it is required.
```

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+- arrow datasets also allow us to filter the data on read which means we don't need to read the full dataset in should only a subset be required.
+
+- arrow datasets work for single files also so we can use the same function to read both single and partitioned target data.
```
Again, my sense is that the advantage here is less about arrow and more about compatibility with the use of arrows in our other hubverse tools like hubEvals?

Arrow is great, but I think the overall proposal is stronger if when framed as something that's compatible with the hubverse while also flexible for admins (i.e., avoid over-emphasizing a specific tool)

annakrystalli replied:
For completeness, adding here that while the proposal has been amended to not be prescriptive on the tooling to write partitioned datasets, it does involve  specific proposal of tooling to access partitioned target data

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+### Functionality for writing, reading and validating target data
+
+As aforementioned, we propose that arrow datasets are used for accessing target data. 
```
I don't think we should state that arrow datasets are a requirement for target data formats (see note above about outlining the desired end state while being agnostic about implementation)

annakrystalli replied:
Hopefully I've made this clearer by re-iterating the specific goal of the proposal: 
```
As aforementioned, we propose that we access target data as arrow datasets using `arrow::open_dataset()`. 
```

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+As aforementioned, we propose that arrow datasets are used for accessing target data. 
+
+We can then develop functionality to:
```
Ah, the points below are exactly what I meant when making the earlier comment re: desired end state (sorry for the thrashing around). Any tool can be used to create target data as long as the criteria below are can be met/implemented in our hubverse packages?

annakrystalli replied:
I've gone ahead and made it a bit clearer that the goal of the proposal is to determine a convention that will allow us to use `open_dataset()` to access target data as arrow datasets. Hopefully it's more explicitly discussed throughout to make that clear.

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+### Other Options Considered
+
+The possibility of non-hive partitioned target data was considered but it was assessed that this would require to much additional configuration to successfully read partitioned data so to begin with, hive partitioning will be sufficient for our needs.
```
What does "additional configuration" mean here? Additional config for hubverse devs who write packages using target data? Additional config for hub admins who create the data?

zkamvar replied:
I also believe an additional option was to require only a single csv or parquet file, but this option could result in large `oracle-output` files for long-running hubs.

elray1 replied:
> I also believe an additional option was to require only a single csv or parquet file, but this option could result in large `oracle-output` files for long-running hubs.

I think the proposal is to allow this as an option in addition to hive-partitioned data, as on line 65 above.  That is, hubs could choose to either use a single csv or parquet file or to use hive partitioning.

annakrystalli replied:
Correct @zkamvar and @elray1 .

To re-iterate:
- we will allow single files which are much simpler to work with  should their size permit
- if we are not using hive style partitioning, the column name for each partition level would need to be provided manually

zkamvar replied:
My suggestion for clarification based on the discussion above.

```suggestion
The possibility of [Directory partitioned](https://arrow.apache.org/docs/r/reference/Partitioning.html) target data was considered but this would require too much additional configuration (i.e. explicit knowledge of the column name for each partition level) to successfully read Directory partitioned data.
```

annakrystalli replied:
Fixed in a7ee5afc13aedaf93ee98b346f71c213a7ab7605

bsweger reviewed (commented):
Thanks, @annakrystalli--this is great and very timely, since we're currently working on target data for variant-nowcast-hub.

I had several comments and questions, mostly trying to parse apart the proposal for file formats and partioning schemes from the many mentions of arrow as a specific tool.

zkamvar commented on decisions/2024-12-09-rfc-record-decisions.md:
```diff
@@ -91,7 +91,7 @@ to establish a standard document to record decisions with the following modifica
 We will use a [project
 poster](https://www.atlassian.com/software/confluence/templates/project-poster)
-for each project with associated artifcats. Each poster describes what the
+for each project with associated artifacts. Each poster describes what the
```
🙈

zkamvar reviewed (commented):
Thank you for writing this up! I really appreciate the examples and links to existing issues that we can work on once we adopt this. It matches three of the four guiding principles (clear, general, and format-agnostic) I outlined in https://github.com/hubverse-org/hubUtils/issues/197#issuecomment-2489141023. 

In general, I agree with @bsweger's assessment that we should focus on the outputs and not the tools. 

I have two questions:

1. Does this proposal exclude using a single file for these target data? 
2. Is there an alternative to prescribing file names? I noticed that we prescribed `timeseries.[csv|parquet]` and `oracle-output.[csv|parquet]`, but this would require existing hubs to adapt their data to this naming scheme (e.g. [FluSight uses `target-hospital-admissions.csv`](https://github.com/cdcepi/FluSight-forecast-hub/blob/main/target-data/target-hospital-admissions.csv) and even the example hub uses `time-series.csv`).

elray1 commented on decisions/2025-01-06-rfc-target-data-formats.md:
````diff
@@ -0,0 +1,95 @@
+
+```
+target-data/oracle-output/target_end_date=2022-10-22
+target-data/oracle-output/target_end_date=2022-10-22/part-0.csv
````
I think it would be helpful to make a clear statement about what, if anything, we're requiring for file names: is it required, recommended, or neither to use a file name like `part-0.csv` or `*-{n}.{csv | parquet}`?

annakrystalli replied:
I've done some digging an added an additional section in the RFC and supplementary material about this topic. In short, file names do not matter as no metadata is extracted from them and ad hoc file names can be accomodated.

bsweger replied:
I suggested some wording above to make this point clearer.

annakrystalli replied:
Fixed in da445fbe819d30c8710dadcedfde1f3608d79d8e

annakrystalli commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,95 @@
+
+- we do not need to be prescriptive about the partitioning scheme used as long as hive partitioning is used. This allows hub admins to partition the data in a way that makes sense for their use case but for hubverse to be able to use open_dataset function to access the data without additional need for defining the partitioning scheme. (see [supplementary materials](https://github.com/reichlab/dec ...
+
+- arrow datasets support both `csv` and `parquet` formats and so we can support both formats for partitioned target data.
```
In the proposal I am suggesting that we (i.e. hubverse tooling) use `arrow` to read the data. So while it's not necessary to use arrow to create target data, compatibility with `arrow` is critical to the proposal for reading.

annakrystalli commented on decisions/2025-01-06-rfc-target-data-formats.md:
````diff
@@ -0,0 +1,95 @@
+
+In Apache Hive, the filename format of partitioned data depends on the partition column names and their values. The files corresponding to each partition are stored in subdirectories, and the directory names encode the partition column names and their values, e.g. `<partition_column_1>=<value_1>/<partition_column_2>=<value_2>/.../<data_files>`
+
+Here's an example of an oracle data partitioned by `target_end_date`:
+
+```
+target-data/oracle-output/target_end_date=2022-10-22
+target-data/oracle-output/target_end_date=2022-10-22/part-0.csv
+target-data/oracle-output/target_end_date=2022-10-29
+target-data/oracle-output/target_end_date=2022-10-29/part-0.csv
+target-data/oracle-output/target_end_date=2022-11-05
+target-data/oracle-output/target_end_date=2022-11-05/part-0.csv
+```
````
This is the output of `fs::dir_ls()` and includes directory names.

nickreich replied:
It seems reasonable to leave the directories in, but I also sympathize with the confusion of having the six lines and three files. It wasn't obvious to me until looking really closely that three were directories and three were files. So maybe we could keep the original list, but add text like @zkamvar has added +/- additional context.

zkamvar replied:
I've updated my suggestion to re-add the directory names based on @nickreich's comment.

annakrystalli replied:
I've amended to use dir_tree instead in e5694aac4064cd728ae820b389f430bd59f9abf4 and f7ede4c01fa13d99892c78d33e975f152d2d2827

annakrystalli commented:
Thanks for the comments @bsweger , @zkamvar  & @elray1 

I've done some additional digging and also responded to your comments in the RFC. More specifically:
- I've clarified that single files are acceptable (I had indeed neglected to make that explicit statement clearly at the beginning of the RFC)
- I clarified that the tools used to create the required structure are not prescribed but mention the arrow tooling as a recommendation
- I have made clearer that the goal of the proposal is to enable **accessing** the data using arrow as an arrow dataset
- I added  sections to both the supplementary materials and the RFC on:
  -  applying a schema to target data datasets
  - ways of splitting data into files that include all columns but can still be accessed as an arrow dataset
  - more detail on file naming and an example of splitting up files that do not use default huve filenames. This may actually be closer to how actual hubs might update their target data.

It would be really useful to get feedback on any assumptions I've made on how actual hubs might update their target data. Such understanding would certainly help evaluating the proposal for fit for purpose!

nickreich commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+Target data files should be either in `csv` or `parquet` format.
+
+Single files for target data are allowed and should be named `timeseries.csv/timeseries.parquet` or `oracle-output.csv|oracle-output.parquet` respectively.
```
```suggestion
Single files for target data are allowed and should be named `timeseries.csv` or `timeseries.parquet` and `oracle-output.csv` or `oracle-output.parquet` respectively.
```

annakrystalli replied:
Removed all mention of csv so suggestion not required

nickreich reviewed (approved):
I think this looks very good and is super thorough! 

One minor comment (that should not preclude merging necessarily) is that I found it confusing to have all these files named `2025-01-06-rfc-target-data-formats`. Maybe the files in the directory named `2025-01-06-rfc-target-data-formats` could have "supplement" added on to their name, so it's clear that they are a supplement, and not the source file for the file in `decisions/`.

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+- `timeseries` format
+- `oracle output` format
+
+As such we will expect data associated with each format to be named either `time-series` or `oracle-output` respectively.
```
```suggestion
As such we will expect data to be stored in `time-series` or `oracle-output` directories within `target-data`.
```

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+- `timeseries` format
+- `oracle output` format
+
+As such we will expect data associated with each format to be named either `time-series` or `oracle-output` respectively.
```
This line is confusing because the naming of the files depends on whether or not the target data is partitioned (in the latter case, we're not prescribing a filename)

```suggestion
```

annakrystalli replied:
Did extra restructuring to address this 6b4208f2ec079306538fa225a2970b6f98526ae0

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+In Apache Hive, the filename format of partitioned data depends on the partition column names and their values. The files corresponding to each partition are stored in subdirectories, and the directory names encode the partition column names and their values, e.g. `<partition_column_1>=<value_1>/<partition_column_2>=<value_2>/.../<data_files>`
+
+Here's an example of an oracle data partitioned by `target_end_date`:
```
```suggestion
Here's an example of an oracle data partitioned by `target_end_date`. Note that individual file names are not prescribed by the Hubverse. Hubs can use their own naming convention or can retain the file names that are generated by the library they use for partitioning:
```

annakrystalli replied:
Added more detail in da445fbe819d30c8710dadcedfde1f3608d79d8e

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+Hive partitioning is the default partitioning scheme used by `arrow::write_dataset()` and `pyarrow.dataset.write_dataset` so no additional configuration beyond the columns to partition by is required when writing partitioned datasets using this functions.
+
+Proposing this convention is designed to allow us to **access target data as arrow datasets using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python.**
```
Checking with @matthewcornell: do we know if hubDataPy will be using `pyarrow.dataset.dataset()`? Regardless, I propose sidestepping the question altogether and focusing on requiring compatibility with Hubverse tools.

```suggestion
Proposing this convention is designed to allow compatibility with R and Python Hubverse packages that access target data.
```

annakrystalli replied:
I edited this to not prescribe the use of arrow in `hubDataPy` but to retain the information that the functionality is available should `arrow` be chosen for `hubDataPy`. I thinks there's no need to remove the fact that the functionality is available

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+- we do not need to be prescriptive about the partitioning scheme used as long as hive partitioning is used. This allows hub admins to partition the data in a way that makes sense for their use case but for hubverse to be able to use open_dataset function to access the data without additional need for defining the partitioning scheme. (see [supplementary materials](https://htmlpreview.github.io/? ...
+
+- `arrow::open_dataset()` supports both `csv` and `parquet` formats and so we can support both formats for partitioned target data.
```
```suggestion
- Hubverse tools support both `csv` and `parquet` formats and so we can support both formats for partitioned target data.
```

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+- partitioned datasets allow more efficient access by partition-aware data readers (for example, arrow, polars, DuckDB), which can filter the data on read if only a subset of it is required.
+
+- `arrow::open_dataset()` works for single files also so we can use the same function to read both single and partitioned target data.
```
```suggestion
- Existing Hubverse tools work for single files also so we can use the same function to read both single and partitioned target data.
```

annakrystalli replied:
Given I'm specifically proposing using `open_dataset()` I think it's fine to specify that it is capable of handling both the formats proposed

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+We can then develop functionality to:
+
+- read target data using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python. https://github.com/hubverse-org/hubUtils/issues/196 
```
```suggestion
- read target data using `arrow::open_dataset()` in R (the Python version of `hubData` is under development and will be compatible with the data format decision we make here)
```
Being clear about the requirement for `arrow::open_dataset()` to work makes more sense here in this list of more technical requirements 👍 

Re: Python...since hubDataPy is still a work in progress, I'm not sure it makes sense to be prescriptive about the function that it will use to read data. It's probably the other way around...hubDataPy will need to accommodate whatever decision we make in this RFC.

zkamvar replied:
@bsweger, could we keep the linked issue? I found those helpful.

elray1 reviewed (approved):
approve modulo Becky's comments r.e. describing formats rather than specific tools, which i support

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+I performed an exploration of applying schema to target data using `arrow::open_dataset()` in the [supplementary materials](https://htmlpreview.github.io/?https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html#applying-a-schema).
+
+In summary, 
```
Thanks for these detailed notes--I'd forgotten how annoying and error-prone the .csv handling can be. Additionally, @matthewcornell found that at least one of the Python tools he was considering (Polars) is not able to do a lazy read of partitioned .csv files. ([related Slack thread](https://reichlab.slack.com/archives/D06D4KUGMC3/p1729110170208309)).

Does it make sense to implement target data partitioning for parquet only? That would get us out of coding around the .csv-related edge cases. If there's demand, we could add .csv partioning later on as a separate feature. 

I'm operating under the assumption that if a hub's target data files are big enough to require partitioning, there's a good chance they're already using parquet.

zkamvar replied:
> Does it make sense to implement target data partitioning for parquet only? That would get us out of coding around the .csv-related edge cases. If there's demand, we could add .csv partioning later on as a separate feature.

I would agree with this assessment. If we decide to implement CSV partitioning at a later date, we can add an LDR.

nickreich replied:
+1 to only support partitioning for parquet format data.

bsweger commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+  - There have generally been issues (see <https://github.com/apache/arrow/issues/34589> and <https://github.com/apache/arrow/issues/34640>) with applying schema to partitioning columns in CSV datasets. Although they have been closed as resolved,
+I [re-commented on one of the issues](https://github.com/apache/arrow/issues/34640#issuecomment-2577363842) as I don't think that is the case. Will be interesting to hear their response.
+
+### Splitting into multiple files that retained all columns
```
This is an interesting technical exploration! That said, since it's up to users to create their target data, as long as `arrow::open_dataset()` can read it, it shouldn't matter of a hub decides to retain the partitioned columns, right? 

Whether or not arrow can write data that includes the partitioning fields as columns isn't really relevant to this RFC, since we're focusing on data access (as you pointed out--great way to frame this!)

annakrystalli replied:
I would rather keep all the information gleaned from the exploration here for the following reason:
- As the expectation is for hub admins to create appropriately formatted data without much help from hubverse functionality, to ensure they succeed means we'll have to write detailed and complete documentation to guide them on what will and what won't work. As such I see all the information in the RFC as useful to form the basis of such documentation. I imagine once the RFC is complete an LDR will be created summarising the decision. I think that document can be slim, without containing all this detailed exploration while retaining all the information here that help with developing documentation. Overall I feel the information is useful and shouldn't be discarded.

bsweger reviewed (commented):
Love how this is evolving, especially the focus on data access!

I made a few wording suggestions to take or leave; the things I feel strongly about are:

- removing the references to specific Python file access methods, since I don't think that's been decided yet
- removing the section about writing target data files with extra columns (because that's a data creation concern, and we've already stated the hubs are responsible for creating target data via whatever method they choose)

And I'll surface this idea here (it's already in a comment):

What if this first iteration states that partitioned target data files must be in parquet format? I know we've already written a lot of code to handle the .csv format for model-output files, so you're a better judge of whether or not that can be re-used here.

But it's worth considering that we don't need to do everything in the first implementation. If limiting partitioned files to .parquet speeds up implementation/reduces amount of code and tests, we can do that and see if there's demand for partitioned .csv files.

bsweger commented:
Just filed a variant-nowcast-hub bug thanks to this RFC and its focus on what's required for successful data access 😄 
https://github.com/reichlab/variant-nowcast-hub/issues/265

zkamvar reviewed (commented):
Thank you for incorporating the changes and the absolutely heroic effort that it took to investigate the CSV partition chaos 😱 💪🏽 

I left some comments and updated my suggestions for clarity. I agree with @bsweger's assessment, specifically on two points: 
1. we should focus less on the specific tools and more on the expected formats.
2. for partitioned data, we should start with parquet (non partitioned can be anything). The value add for partitioned CSV is so small and we don't have any evidence of hubs that have partitioned CSV data.

annakrystalli commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+Proposing this convention is designed to allow us to **access target data as arrow datasets using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python.**
+
+In return the benefit of using this approach are:
```
Fixed in 337574876e04058d9d727f8bb3032644a8ca60d7

annakrystalli commented on decisions/2025-01-06-rfc-target-data-formats.md:
```diff
@@ -0,0 +1,140 @@
+
+- `arrow::open_dataset()` works for single files also so we can use the same function to read both single and partitioned target data.
+
+Large un-partitioned data could be stored locally in a `target-data-raw` directory (which is not committed?) and then written and partitioned into the `target-data` directory (once it has been validated and correct schema is set? see below).
```
I would prefer to get feedback from hub administrators before removing. It might be a tip that helps them think how they will manage the process of partitioning and updating their target data.

annakrystalli replied:
I did go ahead and remove this

annakrystalli commented:
Thanks everyone for the comments.

I've addressed the comments individually but just wanted to add a couple overall responses:
1. I would be very happy with only supporting parquet to start off with. However I feel we should get some feedback from a few hub administrators themselves to ensure we are not being to prescriptive. In any case, I have for now removed mention of `csv` as an accepted file format and moved it to the `Other Options Considered` section, along with the reasoning for rejecting it as an acceptable format for the time being. I feel we should get some approvals from the community before accepting it as it would mean the flusight hub for example is already violating this requirement. (Having said that, they will already need to  change the filename to `oracle-output`).
2. I want to be clear that I am making a **specific and actionable proposal on tooling to access the data**. I have kept mention of tooling for writing data as well as python versions of functions just for informational purposes without making their use prescriptive. However I am proposing the use of **arrow::open_dataset()`** in soon to be developed hubverse functionality  so I don't feel it needs to be removed from the section where I'm explicitly listing the benefits of using it. In my view the proposal has described the format we are expecting and then I'm making a case for how a specific tool can handle the required format. 
3. There was an initial comment by @zkamvar that I missed previously regarding flexibility in filenames (and also pointed out that the example complex forecast hub was using `time-series.csv` instead of `timeseries.csv` which was originally proposed). 
    > Is there an alternative to prescribing file names? I noticed that we prescribed timeseries.[csv|parquet] and oracle-output.[csv|parquet], but this would require existing hubs to adapt their data to this naming scheme (e.g. [FluSight uses target-hospital-admissions.csv](https://github.com/cdcepi/FluSight-forecast-hub/blob/main/target-data/target-hospital-admissions.csv) and even the example hub uses time-series.csv).

    I have corrected all mention in the RFC of `timeseries` to `time-series` but my feeling is that we should just be prescriptive in the filename instead of allowing flexibility. Flexibility would likely require a property in one of the config files which for now we have not mentioned. It may be that at some point we will require information from hub admins to be able to successfully make use of time-series data in the form of config so perhaps we could revisit this when we reach that point?

Overall, because we are not providing tooling for creating target data, this actually means we will need to provide clear documentation for hub admins to ensure they succeed. So detail and specifics will be important to make our documentation helpful to hub admins. This RFC and the experimentation within can be a good source of information for our documentation. I've added a comment regarding documentation in the proposal also 76bfbb136bb5c5bed111c8e3e7bdfb7dc93ca574.

annakrystalli commented:
If folks are happy with the current version, I can share it on more widely to get some input from hub admins as well.

annakrystalli commented:
Thanks everyone for your feedback and patience and helping work through the RFC process too.  Merging!
